src/Router.sol

```diff
@@ -170,28 +172,61 @@ contract Router is ReentrancyGuard, IRouter, ScheduledUpgradeable, AccessControl
     }
 
     /// @notice Relays tokens to the recipient and stores a receipt
-    /// @param token The token being relayed
+    /// @param requestId The original request ID from the source chain
+    /// @param sender The sender of the swap request on the source chain
     /// @param recipient The target recipient of the tokens
+    /// @param tokenIn The token being relayed on the destination chain
+    /// @param tokenOut The output token on the destination chain
     /// @param amountOut The amount transferred to the recipient on the destination chain
-    /// @param requestId The original request ID from the source chain
     /// @param srcChainId The ID of the source chain where the request originated
-    function relayTokens(address token, address recipient, uint256 amountOut, bytes32 requestId, uint256 srcChainId)
-        external
-        nonReentrant
-    {
+    /// @param nonce The nonce used for the swap request
+    function relayTokens(
+        bytes32 requestId,
+        address sender,
+        address recipient,
+        address tokenIn,
+        address tokenOut,
+        uint256 amountOut,
+        uint256 srcChainId,
+        uint256 nonce
+    ) external nonReentrant {
         require(!swapRequestReceipts[requestId].fulfilled, ErrorsLib.AlreadyFulfilled());
-        require(token != address(0) && recipient != address(0), ErrorsLib.InvalidTokenOrRecipient());
+        require(
+            tokenIn != address(0) && tokenOut != address(0) && sender != address(0) && recipient != address(0),
+            ErrorsLib.InvalidTokenOrRecipient()
+        );
         require(amountOut > 0, ErrorsLib.ZeroAmount());
+        require(
+            srcChainId != getChainID(),
+            ErrorsLib.SourceChainIdShouldBeDifferentFromDestination(srcChainId, getChainID())
+        );
+        require(
+            requestId
+                == keccak256(
+                    abi.encode(
+                        sender,
+                        recipient,
+                        tokenIn,
+                        tokenOut,
+                        amountOut,
+                        srcChainId,
+                        // the relayTokens function is called on the destination chain, so dstChainId is the current chain ID
+                        getChainID(),
+                        nonce
+                    )
+                ),
+            ErrorsLib.SwapRequestParametersMismatch()
+        );
 
         fulfilledTransfers.add(requestId);
 
-        IERC20(token).safeTransferFrom(msg.sender, recipient, amountOut);
+        IERC20(tokenIn).safeTransferFrom(msg.sender, recipient, amountOut);
```

Comment:
shouldn't this be `tokenOut` as we're on the destination chain?

Reply:
looking at the tests, I'm surprised this doesn't get caught. Is it because the unit tests don't have a notion of multiple chains or something?

Reply:
Maybe tests are assuming it's the same value for both everytime?